src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1096 @@
+		const long delay_ms = 1000;
+		(void) WaitLatch(MyLatch,
+						 WL_LATCH_SET | WL_TIMEOUT | WL_EXIT_ON_PM_DEATH,
+						 delay_ms,
+						 WAIT_EVENT_PG_SLEEP);
+
+		ResetLatch(MyLatch);
+	}
+
+	MemoryContextSwitchTo(oldContext);
+	MemoryContextDelete(waitContext);
+
+	PG_RETURN_BOOL(reachedDesiredState);
+}
+
+
+BackgroundWorkerHandle *
+StartCitusBackgroundTaskQueueMonitor(Oid database, Oid extensionOwner)
+{
+	BackgroundWorker worker = { 0 };
+	BackgroundWorkerHandle *handle = NULL;
+
+	/* Configure a worker. */
+	memset(&worker, 0, sizeof(worker));
+	SafeSnprintf(worker.bgw_name, BGW_MAXLEN,
+				 "Citus Background Task Queue Monitor: %u/%u",
+				 database, extensionOwner);
+	worker.bgw_flags =
+		BGWORKER_SHMEM_ACCESS | BGWORKER_BACKEND_DATABASE_CONNECTION;
+	worker.bgw_start_time = BgWorkerStart_ConsistentState;
+
+	/* don't restart, we manage restarts from maintenance daemon */
+	worker.bgw_restart_time = BGW_NEVER_RESTART;
+	strcpy_s(worker.bgw_library_name, sizeof(worker.bgw_library_name), "citus");
+	strcpy_s(worker.bgw_function_name, sizeof(worker.bgw_library_name),
+			 "CitusBackgroundTaskQueueMonitorMain");
+	worker.bgw_main_arg = ObjectIdGetDatum(MyDatabaseId);
+	memcpy_s(worker.bgw_extra, sizeof(worker.bgw_extra), &extensionOwner,
+			 sizeof(Oid));
+	worker.bgw_notify_pid = MyProcPid;
+
+	if (!RegisterDynamicBackgroundWorker(&worker, &handle))
+	{
+		return NULL;
+	}
+
+	pid_t pid;
+	WaitForBackgroundWorkerStartup(handle, &pid);
+
+	return handle;
+}
+
+
+void
+CitusBackgroundTaskQueueMonitorMain(Datum arg)
+{
+	Oid databaseOid = DatumGetObjectId(arg);
+
+	/* extension owner is passed via bgw_extra */
+	Oid extensionOwner = InvalidOid;
+	memcpy_s(&extensionOwner, sizeof(extensionOwner),
+			 MyBgworkerEntry->bgw_extra, sizeof(Oid));
+
+	BackgroundWorkerUnblockSignals();
+
+	/* connect to database, after that we can actually access catalogs */
+	BackgroundWorkerInitializeConnectionByOid(databaseOid, extensionOwner, 0);
+
+	/*
+	 * There should be exactly one background task monitor running, running multiple would
+	 * cause conflicts on processing the tasks in the catalog table as well as violate
+	 * parallelism guarantees. To make sure there is at most, exactly one backend running
+	 * we take a session lock on the CITUS_BACKGROUND_TASK_MONITOR operation.
+	 *
+	 * TODO now that we have a lock, we should install a term handler to terminate any
+	 * 'child' backend when we are terminated. Otherwise we will still have a situation
+	 * where the actual task could be running multiple times.
+	 */
+	LOCKTAG tag = { 0 };
+	SET_LOCKTAG_CITUS_OPERATION(tag, CITUS_BACKGROUND_TASK_MONITOR);
+	const bool sessionLock = true;
+	const bool dontWait = true;
+	LockAcquireResult locked =
+		LockAcquire(&tag, AccessExclusiveLock, sessionLock, dontWait);
+	if (locked == LOCKACQUIRE_NOT_AVAIL)
+	{
+		ereport(ERROR, (errmsg("background task queue monitor already running for "
+							   "database")));
+		exit(0);
+	}
+
```

Comment:
it might be good to also have similar type of locking on the task ID, such that if the task ID is still running, we can recognize that by attempting to acquire the lock and kill the PID if we cannot acquire the lock.

that does require passing the task ID to the process

when we do this, it might be nice to include task ID in appname